### PR TITLE
[MODINREACH-174] INN-Reach Circulation Flow: React to Change in Request…

### PR DESCRIPTION
**Pull request by aleksandr2110**
… Record for INN-Reach Transaction: Cancel Request (Patron Hold)

## Purpose
https://issues.folio.org/browse/MODINREACH-174

## Approach
1. mod-inn-reach should consume request-storage domain events via Kafka ([CIRCSTORE-302](https://issues.folio.org/browse/CIRCSTORE-302))
2. mod-inn-reach should check if the request is associated with an active INN-Reach Patron Hold transaction
3. If so the then mod-inn-reach should issue a call to the cancel item hold endpoint (D2IR API reference, pg. 54) and set the patron hold transaction status to BORROWING_SITE_CANCEL

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 1]
```diff
@@ -274,6 +261,41 @@ private PatronHoldCheckInResponseDTO checkInItem(InnReachTransaction transaction
...
+    var requestId = requestDTO.getId();
+    var itemId = requestDTO.getItemId();
+    if (requestDTO.getStatus() == CLOSED_CANCELLED) {
+      var transactionItemHold = (TransactionItemHold) transaction.getHold();
+      var instance = instanceStorageClient.getInstanceById(requestDTO.getInstanceId());
+      transaction.setState(CANCEL_REQUEST);
+
+      reportOwningSiteCancel(transaction, instance.getHrid(), transactionItemHold.getPatronName());
+      return;
+    }
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: any logs?

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 2]
```diff
@@ -274,6 +261,41 @@ private PatronHoldCheckInResponseDTO checkInItem(InnReachTransaction transaction
...
+      reportTransferRequest(transaction, item.getHrid());
+    }
+  }
+
+  private void requestAssociatedWithPatronHoldTransaction(RequestDTO requestDTO, InnReachTransaction transaction) {
+    if (requestDTO.getStatus() == CLOSED_CANCELLED) {
+      transaction.setState(BORROWING_SITE_CANCEL);
+      reportCancelItemHold(transaction);
+    }
+  }
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: any logs?

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java [hunk 3]
```diff
@@ -314,6 +316,28 @@ void shouldUpdateTransactionWhenCancelRequest() {
...
+    "classpath:db/central-server/pre-populate-central-server.sql",
+    "classpath:db/inn-reach-transaction/pre-populate-another-inn-reach-transaction.sql"
+  })
+  void shouldUpdateTransactionWithPatronHoldWhenCancelRequest() {
+    var event = getRequestDomainEvent(DomainEventType.UPDATED);
+    var request = event.getData().getNewEntity();
+    request.setId(PRE_POPULATED_PATRON_TRANSACTION_REQUEST_ID);
+    request.setStatus(CLOSED_CANCELLED);
+
+    when(innReachExternalService.postInnReachApi(any(), any())).thenReturn("ok");
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: unnecessary mock

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 4]
```diff
@@ -198,36 +200,21 @@ public void handleLoanUpdate(LoanDTO loan) {
...
-      hold.setFolioItemId(itemId);
-      hold.setFolioInstanceId(requestDTO.getInstanceId());
-      hold.setFolioHoldingId(requestDTO.getHoldingsRecordId());
-      hold.setFolioItemBarcode(item.getBarcode());
-      transaction.setState(TRANSFER);
-
-      reportTransferRequest(transaction, item.getHrid());
+    switch (transaction.getType()) {
+      case ITEM:
+        requestAssociatedWithItemHoldTransaction(requestDTO, transaction);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: rename it something more meaningful. The method name should state the action they perform

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 5]
```diff
@@ -198,36 +200,21 @@ public void handleLoanUpdate(LoanDTO loan) {
...
+        break;
+      case PATRON:
+        requestAssociatedWithPatronHoldTransaction(requestDTO, transaction);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: same as above

Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 4 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 2 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 1 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Fixed

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java (hunk 3 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: Fixed

**Review by oliinyko — APPROVED** [review]

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java [hunk 6]
```diff
@@ -314,6 +316,26 @@ void shouldUpdateTransactionWhenCancelRequest() {
...
+    "classpath:db/central-server/pre-populate-central-server.sql",
+    "classpath:db/inn-reach-transaction/pre-populate-another-inn-reach-transaction.sql"
+  })
+  void shouldUpdateTransactionWithPatronHoldWhenCancelRequest() {
+    var event = getRequestDomainEvent(DomainEventType.UPDATED);
+    var request = event.getData().getNewEntity();
+    request.setId(PRE_POPULATED_PATRON_TRANSACTION_REQUEST_ID);
+    request.setStatus(CLOSED_CANCELLED);
+
+    listener.handleRequestEvents(asSingleConsumerRecord(CIRC_REQUEST_TOPIC, REQUEST_ID, event));
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: use same id for event key as it's a bit confusing why they are different

Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: Fixed

**Pull request merged**